python/ray/serve/batching.py

```diff
@@ -137,7 +137,15 @@ async def wait_for_batch(self) -> List[Any]:
             # Add all new arrivals to the batch.
             while len(batch) < max_batch_size and not self.queue.empty():
                 batch.append(self.queue.get_nowait())
-            self.queue_put_event.clear()
+            
+            # Only clear the put event if the queue is empty. If it's not empty
+            # we can start construcing a new batch immediately in the next loop.
+            # The code that puts items into the queue runs on the same event loop
+            # as this code, so there's no race condition between the time we
+            # get objects in the queue (and clear the event) and when objects
+            # get added to the queue.
+            if self.queue.empty():
+                self.queue_put_event.clear()
```

Comment:
These are the lines relevant to the bug.

We used to clear the `queue_put_event` whenever a batch was constructed. This meant that even if the request queue had enough requests to create another batch immediately, the batch handler would wait until another request was sent before constructing a batch. If no request was sent, the existing requests waited indefinitely.

Now, we clear the event only if the queue is empty.

Reply:
Actually, the pending requests don't wait indefinitely. They wait until the `batch_wait_timeout_s` is hit.